### PR TITLE
Make User#update more typesafe

### PR DESCRIPTION
**Pull request by whalemer**
Would let us catch people mistyping `user.update('blockpms')` and the like, if any of those files were typescripted.

### CODE REVIEW

**Comment by TheJetOU** [comment]
Is it actually `keyof User` though? Because `SETTINGS` doesn't contain all of the keys in `User`. Something such as `user.update('connections')` also doesn't look correct.

I would think it would be something like:
```typescript
const SETTINGS = ['isSysop', ...] as const;

function getUpdateuserText(...updated: typeof SETTINGS) { ... }
```

File: server/users.ts [hunk 1]
```diff
@@ -1018,20 +1018,22 @@ export class User extends Chat.MessageContext {
...
-			// @ts-ignore - dynamic lookup
 			diff[setting] = this[setting];
 		}
 		return `|updateuser|${this.getIdentityWithStatus()}|${named}|${this.avatar}|${JSON.stringify(diff)}`;
 	}
 	/**
 	 * @param updated the settings which have been updated or none for all settings.
 	 */
-	update(...updated: string[]) {
+	update(...updated: (keyof User)[]) {
```
Comment (Zarel) [inline server/users.ts]: Wait, why does `update` support partial updates in the first place? Did I approve this?

Comment (Zarel) [reply server/users.ts]: apparently it was #5438

Comment (Zarel) [reply server/users.ts]: `keyof User` seems way too broad, anyway; it should be a subset of `SETTINGS`.

I would do `SETTINGS = [...] as const`, and then `typeof SETTINGS[0]` or something?

**Comment by whalemer** [comment]
Had no idea about `as const`, looks like a nice feature.

File: server/users.ts [hunk 2]
```diff
@@ -1018,20 +1019,20 @@ export class User extends Chat.MessageContext {
 	/**
 	 * @param updated the settings which have been updated or none for all settings.
 	 */
-	getUpdateuserText(...updated: string[]) {
+	getUpdateuserText(...updated: UserSetting[]) {
 		const named = this.named ? 1 : 0;
-		const diff = {};
+		const diff: Partial<Pick<User, UserSetting>> = {};
```
Comment (Zarel) [inline server/users.ts]: tbh I would just let `diff` be `any`. It's about to become JSON anyway, and building JSON usually requires things TS doesn't understand.

**Comment by Zarel** [comment]
Cool, thanks!

**Pull request merged**